README.rst
pylibdmtxp


CHANGELOG.md
###

### v0.1.10

* #68 Add support for images loaded using imageio
* #67 Python 3.8., 3.9 and 3.10; drop support for Python 3.4
* #41 Alter encoding example to show utf-8 encoding
* #39 Fix Travis / Python 3.7


### v0.1.9

* #36 Fix Windows build

### v0.1.8

* #34 Python 3.7
* #24 Support libdmtx 0.7.5 (fixes decoding error)
* #22 Use reStructuredText for README
* #18 Encode support
* #16 Improve error messages

### v0.1.6

* #11 Python 3.6

### v0.1.6

* #9 Check for empty decoded value

### v0.1.5

* #7 Support older numpy

### v0.1.4

* #5 Better handling of DLLs

### v0.1.3

* #3 Convert images only as required

### v0.1.2

* #1 Incorrect handling of bits-per-pixels

### v0.1.1

* Long description onto PyPI

### v0.1.0

* Initial release


pylibdmtx/tests/test_pylibdmtx.py
import unittest

from pathlib import Path

try:
    from unittest.mock import call, patch
except ImportError:
    # Python 2
    from mock import call, patch

import numpy as np

from PIL import Image

try:
    import cv2
except ImportError:
    cv2 = None

try:
    import imageio
except ImportError:
    imageio = None

from pylibdmtx.pylibdmtx import (
    decode, encode, Decoded, Encoded, Rect, EXTERNAL_DEPENDENCIES
)
from pylibdmtx.pylibdmtx_error import PyLibDMTXError


TESTDATA = Path(__file__).parent

@unittest.skip("skipping decode test")
class TestDecode(unittest.TestCase):
    EXPECTED = [
        Decoded(
            data=b'Stegosaurus',
            rect=Rect(left=5, top=6, width=96, height=95)
        ),
        Decoded(
            data=b'Plesiosaurus',
            rect=Rect(left=298, top=6, width=95, height=95)
        )
    ]

    @classmethod
    def setUpClass(cls):
        cls.datamatrix = Image.open(str(TESTDATA.joinpath('datamatrix.png')))
        cls.empty = Image.open(str(TESTDATA.joinpath('empty.png')))

        # assertRaisesRegexp was a deprecated alias removed in Python 3.11
        if not hasattr(cls, 'assertRaisesRegex'):
            cls.assertRaisesRegex = cls.assertRaisesRegexp

    @classmethod
    def tearDownClass(cls):
        cls.datamatrix = cls.empty = None

    def test_decode(self):
        "Read both barcodes in `datamatrix.png`"
        res = decode(self.datamatrix)
        self.assertEqual(self.EXPECTED, res)

    def test_decode_single(self):
        "Read just one of the barcodes in `datamatrix.png`"
        res = decode(self.datamatrix, max_count=1)
        self.assertEqual(self.EXPECTED[:1], res)

    def test_decode_tuple(self):
        "Read barcodes in pixels"
        pixels = self.datamatrix.copy().convert('RGB').tobytes()
        width, height = self.datamatrix.size[:2]
        res = decode((pixels, width, height))
        self.assertEqual(self.EXPECTED, res)

    def test_empty(self):
        "Do not show any output for an image that does not contain a barcode"
        res = decode(self.empty)
        self.assertEqual([], res)

    def test_decode_numpy(self):
        "Read image using Pillow and convert to numpy.ndarray"
        res = decode(np.asarray(self.datamatrix))
        self.assertEqual(self.EXPECTED, res)

    @unittest.skipIf(imageio is None, 'imageio not installed')
    def test_decode_imageio(self):
        "Read image using imageio"
        res = decode(imageio.imread(TESTDATA.joinpath('datamatrix.png')))
        self.assertEqual(self.EXPECTED, res)

    @unittest.skipIf(cv2 is None, 'OpenCV not installed')
    def test_decode_opencv(self):
        "Read image using OpenCV"
        res = decode(cv2.imread(str(TESTDATA.joinpath('datamatrix.png'))))
        self.assertEqual(self.EXPECTED, res)

    def test_external_dependencies(self):
        "External dependencies"
        self.assertEqual(1, len(EXTERNAL_DEPENDENCIES))
        self.assertIn('libdmtx', EXTERNAL_DEPENDENCIES[0]._name)

    @patch('pylibdmtx.pylibdmtx.dmtxImageCreate')
    def test_dmtxImageCreate_failed(self, dmtxImageCreate):
        dmtxImageCreate.return_value = None
        self.assertRaisesRegex(
            PyLibDMTXError, 'Could not create image', decode, self.datamatrix
        )
        self.assertEqual(1, dmtxImageCreate.call_count)

    @patch('pylibdmtx.pylibdmtx.dmtxDecodeCreate')
    def test_dmtxDecodeCreate_failed(self, dmtxDecodeCreate):
        dmtxDecodeCreate.return_value = None
        self.assertRaisesRegex(
            PyLibDMTXError, 'Could not create decoder', decode, self.datamatrix
        )
        self.assertEqual(1, dmtxDecodeCreate.call_count)

    def test_unsupported_bits_per_pixel(self):
        # 40 bits-per-pixel
        data = (list(range(3 * 3 * 5)), 3, 3)
        self.assertRaisesRegex(
            PyLibDMTXError,
            (
                r'Unsupported bits-per-pixel: \[40\] Should be one of '
                r'\[8, 16, 24, 32\]'
            ),
            decode, data
        )

    def test_inconsistent_dimensions(self):
        # Image data has ten bytes. width x height indicates nine bytes
        data = (list(range(10)), 3, 3)
        self.assertRaisesRegex(
            PyLibDMTXError,
            (
                r'Inconsistent dimensions: image data of 10 bytes is not '
                r'divisible by \(width x height = 9\)'
            ),
            decode, data
        )


@unittest.skip("skipping encode test")
class TestEncode(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        # assertRaisesRegexp was a deprecated alias removed in Python 3.11
        if not hasattr(cls, 'assertRaisesRegex'):
            cls.assertRaisesRegex = cls.assertRaisesRegexp

    def _assert_encoded_data(self, expected_data, encoded):
        # Check encoded data
        image = Image.frombytes(
            'RGB', (encoded.width, encoded.height), encoded.pixels
        )
        decoded = decode(image)

        self.assertEqual(1, len(decoded))
        self.assertEqual(expected_data, decoded[0].data)

    def test_encode_defaults(self):
        data = b'hello world'
        encoded = encode(data)

        # Check returned data, with the exception of the pixel data
        self.assertEqual(
            Encoded(width=100, height=100, bpp=24, pixels=None),
            encoded._replace(pixels=None)
        )
        self._assert_encoded_data(data, encoded)

    def test_encode_size(self):
        data = b'hello world'
        encoded = encode(data, size='36x36')

        # Check returned data, with the exception of the pixel data
        self.assertEqual(
            Encoded(width=200, height=200, bpp=24, pixels=None),
            encoded._replace(pixels=None)
        )

        self._assert_encoded_data(data, encoded)

    def test_encode_scheme(self):
        data = b'hello world'
        encoded = encode(data, scheme='Base256')

        # Check returned data, with the exception of the pixel data
        self.assertEqual(
            Encoded(width=110, height=110, bpp=24, pixels=None),
            encoded._replace(pixels=None)
        )

        self._assert_encoded_data(data, encoded)

    def test_invalid_scheme(self):
        self.assertRaisesRegex(
            PyLibDMTXError,
            r"Invalid scheme \[asdf\]: should be one of \['Ascii",
            encode, b' ', scheme='asdf'
        )

    def test_invalid_size(self):
        self.assertRaisesRegex(
            PyLibDMTXError,
            r"Invalid size \[9x9\]: should be one of \['RectAuto'",
            encode, b' ', size='9x9'
        )

    def test_image_not_large_enough(self):
        self.assertRaisesRegex(
            PyLibDMTXError,
            'Could not encode data, possibly because the image is not '
            'large enough to contain the data',
            encode, b' '*50, size='10x10'
        )


if __name__ == '__main__':
    unittest.main()


pylibdmtxp/tests/test_pylibdmtx.py
import unittest

from pathlib import Path

try:
    from unittest.mock import call, patch
except ImportError:
    # Python 2
    from mock import call, patch

import numpy as np

from PIL import Image

try:
    import cv2
except ImportError:
    cv2 = None

try:
    import imageio
except ImportError:
    imageio = None

from pylibdmtxp.pylibdmtx import (
    decode, encode, Decoded, Encoded, Rect, EXTERNAL_DEPENDENCIES
)
from pylibdmtxp.pylibdmtx_error import PyLibDMTXError


TESTDATA = Path(__file__).parent

@unittest.skip("skipping decode test")
class TestDecode(unittest.TestCase):
    EXPECTED = [
        Decoded(
            data=b'Stegosaurus',
            rect=Rect(left=5, top=6, width=96, height=95)
        ),
        Decoded(
            data=b'Plesiosaurus',
            rect=Rect(left=298, top=6, width=95, height=95)
        )
    ]

    @classmethod
    def setUpClass(cls):
        cls.datamatrix = Image.open(str(TESTDATA.joinpath('datamatrix.png')))
        cls.empty = Image.open(str(TESTDATA.joinpath('empty.png')))

        # assertRaisesRegexp was a deprecated alias removed in Python 3.11
        if not hasattr(cls, 'assertRaisesRegex'):
            cls.assertRaisesRegex = cls.assertRaisesRegexp

    @classmethod
    def tearDownClass(cls):
        cls.datamatrix = cls.empty = None

    def test_decode(self):
        "Read both barcodes in `datamatrix.png`"
        res = decode(self.datamatrix)
        self.assertEqual(self.EXPECTED, res)

    def test_decode_single(self):
        "Read just one of the barcodes in `datamatrix.png`"
        res = decode(self.datamatrix, max_count=1)
        self.assertEqual(self.EXPECTED[:1], res)

    def test_decode_tuple(self):
        "Read barcodes in pixels"
        pixels = self.datamatrix.copy().convert('RGB').tobytes()
        width, height = self.datamatrix.size[:2]
        res = decode((pixels, width, height))
        self.assertEqual(self.EXPECTED, res)

    def test_empty(self):
        "Do not show any output for an image that does not contain a barcode"
        res = decode(self.empty)
        self.assertEqual([], res)

    def test_decode_numpy(self):
        "Read image using Pillow and convert to numpy.ndarray"
        res = decode(np.asarray(self.datamatrix))
        self.assertEqual(self.EXPECTED, res)

    @unittest.skipIf(imageio is None, 'imageio not installed')
    def test_decode_imageio(self):
        "Read image using imageio"
        res = decode(imageio.imread(TESTDATA.joinpath('datamatrix.png')))
        self.assertEqual(self.EXPECTED, res)

    @unittest.skipIf(cv2 is None, 'OpenCV not installed')
    def test_decode_opencv(self):
        "Read image using OpenCV"
        res = decode(cv2.imread(str(TESTDATA.joinpath('datamatrix.png'))))
        self.assertEqual(self.EXPECTED, res)

    def test_external_dependencies(self):
        "External dependencies"
        self.assertEqual(1, len(EXTERNAL_DEPENDENCIES))
        self.assertIn('libdmtx', EXTERNAL_DEPENDENCIES[0]._name)

    @patch('pylibdmtxp.pylibdmtx.dmtxImageCreate')
    def test_dmtxImageCreate_failed(self, dmtxImageCreate):
        dmtxImageCreate.return_value = None
        self.assertRaisesRegex(
            PyLibDMTXError, 'Could not create image', decode, self.datamatrix
        )
        self.assertEqual(1, dmtxImageCreate.call_count)

    @patch('pylibdmtxp.pylibdmtx.dmtxDecodeCreate')
    def test_dmtxDecodeCreate_failed(self, dmtxDecodeCreate):
        dmtxDecodeCreate.return_value = None
        self.assertRaisesRegex(
            PyLibDMTXError, 'Could not create decoder', decode, self.datamatrix
        )
        self.assertEqual(1, dmtxDecodeCreate.call_count)

    def test_unsupported_bits_per_pixel(self):
        # 40 bits-per-pixel
        data = (list(range(3 * 3 * 5)), 3, 3)
        self.assertRaisesRegex(
            PyLibDMTXError,
            (
                r'Unsupported bits-per-pixel: \[40\] Should be one of '
                r'\[8, 16, 24, 32\]'
            ),
            decode, data
        )

    def test_inconsistent_dimensions(self):
        # Image data has ten bytes. width x height indicates nine bytes
        data = (list(range(10)), 3, 3)
        self.assertRaisesRegex(
            PyLibDMTXError,
            (
                r'Inconsistent dimensions: image data of 10 bytes is not '
                r'divisible by \(width x height = 9\)'
            ),
            decode, data
        )


@unittest.skip("skipping encode test")
class TestEncode(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        # assertRaisesRegexp was a deprecated alias removed in Python 3.11
        if not hasattr(cls, 'assertRaisesRegex'):
            cls.assertRaisesRegex = cls.assertRaisesRegexp

    def _assert_encoded_data(self, expected_data, encoded):
        # Check encoded data
        image = Image.frombytes(
            'RGB', (encoded.width, encoded.height), encoded.pixels
        )
        decoded = decode(image)

        self.assertEqual(1, len(decoded))
        self.assertEqual(expected_data, decoded[0].data)

    def test_encode_defaults(self):
        data = b'hello world'
        encoded = encode(data)

        # Check returned data, with the exception of the pixel data
        self.assertEqual(
            Encoded(width=100, height=100, bpp=24, pixels=None),
            encoded._replace(pixels=None)
        )
        self._assert_encoded_data(data, encoded)

    def test_encode_size(self):
        data = b'hello world'
        encoded = encode(data, size='36x36')

        # Check returned data, with the exception of the pixel data
        self.assertEqual(
            Encoded(width=200, height=200, bpp=24, pixels=None),
            encoded._replace(pixels=None)
        )

        self._assert_encoded_data(data, encoded)

    def test_encode_scheme(self):
        data = b'hello world'
        encoded = encode(data, scheme='Base256')

        # Check returned data, with the exception of the pixel data
        self.assertEqual(
            Encoded(width=110, height=110, bpp=24, pixels=None),
            encoded._replace(pixels=None)
        )

        self._assert_encoded_data(data, encoded)

    def test_invalid_scheme(self):
        self.assertRaisesRegex(
            PyLibDMTXError,
            r"Invalid scheme \[asdf\]: should be one of \['Ascii",
            encode, b' ', scheme='asdf'
        )

    def test_invalid_size(self):
        self.assertRaisesRegex(
            PyLibDMTXError,
            r"Invalid size \[9x9\]: should be one of \['RectAuto'",
            encode, b' ', size='9x9'
        )

    def test_image_not_large_enough(self):
        self.assertRaisesRegex(
            PyLibDMTXError,
            'Could not encode data, possibly because the image is not '
            'large enough to contain the data',
            encode, b' '*50, size='10x10'
        )


if __name__ == '__main__':
    unittest.main()


pylibdmtxp/tests/test_dmtx_library.py
import unittest

from pathlib import Path

try:
    from unittest.mock import call, patch
except ImportError:
    # Python 2
    from mock import call, patch

from pylibdmtxp import dmtx_library


@unittest.skip("skipping test")
class TestLoad(unittest.TestCase):
    def setUp(self):
        self.addCleanup(patch.stopall)
        self.cdll = patch(
            'pylibdmtxp.dmtx_library.cdll', autospec=True
        ).start()
        self.find_library = patch(
            'pylibdmtxp.dmtx_library.find_library', autospec=True
        ).start()
        self.platform = patch(
            'pylibdmtxp.dmtx_library.platform', autospec=True
        ).start()
        self.windows_fname = patch(
            'pylibdmtxp.dmtx_library._windows_fname', autospec=True,
            return_value='dll fname'
        ).start()

    def test_found_non_windows(self):
        "libdmtx loaded ok on non-Windows platform"
        self.platform.system.return_value = 'Not windows'

        res = dmtx_library.load()

        self.platform.system.assert_called_once_with()
        self.find_library.assert_called_once_with('dmtx')
        self.cdll.LoadLibrary.assert_called_once_with(
            self.find_library.return_value
        )

        self.assertEqual(self.cdll.LoadLibrary.return_value, res)
        self.assertEqual(0, self.windows_fname.call_count)

    def test_not_found_non_windows(self):
        "libdmtx not found on non-Windows platform"
        self.platform.system.return_value = 'Not windows'
        self.find_library.return_value = None

        self.assertRaises(ImportError, dmtx_library.load)

        self.platform.system.assert_called_once_with()
        self.find_library.assert_called_once_with('dmtx')

    def test_found_windows(self):
        "libdmtx found on Windows"
        self.platform.system.return_value = 'Windows'

        res = dmtx_library.load()

        self.platform.system.assert_called_once_with()
        self.cdll.LoadLibrary.assert_called_once_with(
            self.windows_fname.return_value
        )
        self.assertEqual(self.cdll.LoadLibrary.return_value, res)

    def test_found_second_attempt_windows(self):
        "libdmtx found on the second attempt on Windows"
        self.platform.system.return_value = 'Windows'
        self.cdll.LoadLibrary.side_effect = [
            OSError,            # First call does not load DLL
            'loaded library',   # Second call loads DLL
        ]

        res = dmtx_library.load()

        self.platform.system.assert_called_once_with()
        self.cdll.LoadLibrary.assert_has_calls([
            call(self.windows_fname.return_value),
            call(str(Path(dmtx_library.__file__).parent.joinpath(
                self.windows_fname.return_value
            ))),
        ])

        self.assertEqual('loaded library', res)

    def test_not_found_windows(self):
        "libdmtx not found on Windows"
        self.platform.system.return_value = 'Windows'
        self.cdll.LoadLibrary.side_effect = OSError

        self.assertRaises(OSError, dmtx_library.load)

        self.platform.system.assert_called_once_with()
        # Two attempts at loading
        self.cdll.LoadLibrary.assert_has_calls([
            call(self.windows_fname.return_value),
            call(str(Path(dmtx_library.__file__).parent.joinpath(
                self.windows_fname.return_value
            ))),
        ])


@unittest.skip("skipping test")
class TestWindowsFname(unittest.TestCase):
    def setUp(self):
        self.addCleanup(patch.stopall)
        self.sys = patch('pylibdmtxp.dmtx_library.sys', autospec=True).start()

    def test_32bit(self):
        self.sys.maxsize = 2**32
        self.assertEqual('libdmtx-32.dll', dmtx_library._windows_fname())

    def test_64bit(self):
        # This is a 'long' on a 32-bit interpreter
        self.sys.maxsize = 2**32 + 1
        self.assertEqual('libdmtx-64.dll', dmtx_library._windows_fname())


if __name__ == '__main__':
    unittest.main()


pylibdmtxp/scripts/write_datamatrix.py
#!/usr/bin/env python
from __future__ import print_function

import argparse
import sys

import pylibdmtxp
from pylibdmtxp.pylibdmtx import (
    encode, ENCODING_SIZE_NAMES, ENCODING_SCHEME_NAMES
)


def main(args=None):
    if args is None:
        args = sys.argv[1:]

    parser = argparse.ArgumentParser(
        description='Writes a datamatrix barcode to a new image file'
    )
    parser.add_argument('file', help='Filename of the output image')
    parser.add_argument(
        'data', help='Data to be written; will be utf-8 encoded'
    )
    parser.add_argument(
        '--size',
        help="Encoding size (not image dimensions); default is 'ShapeAuto'",
        choices=ENCODING_SIZE_NAMES
    )
    parser.add_argument(
        '--scheme',
        help="Encoding method; default is 'Ascii'",
        choices=ENCODING_SCHEME_NAMES
    )
    parser.add_argument(
        '-v', '--version', action='version',
        version='%(prog)s ' + pylibdmtx.__version__
    )
    args = parser.parse_args(args)

    from PIL import Image

    encoded = encode(
        args.data.encode('utf-8'), size=args.size, scheme=args.scheme
    )
    im = Image.frombytes('RGB', (encoded.width, encoded.height), encoded.pixels)
    im.save(args.file)


if __name__ == '__main__':
    main()


pylibdmtx/pylibdmtx.py
from __future__ import print_function

import ctypes
from collections import namedtuple
from contextlib import contextmanager
from ctypes import byref, cast, string_at
from functools import partial

from .pylibdmtx_error import PyLibDMTXError
from .wrapper import (
    c_ubyte_p, dmtxImageCreate, dmtxImageDestroy, dmtxDecodeCreate,
    dmtxDecodeDestroy, dmtxRegionDestroy, dmtxMessageDestroy, dmtxTimeAdd,
    dmtxTimeNow, dmtxDecodeMatrixRegion, dmtxRegionFindNext,
    dmtxMatrix3VMultiplyBy, dmtxDecodeSetProp, DmtxPackOrder, DmtxProperty,
    DmtxUndefined, DmtxVector2, EXTERNAL_DEPENDENCIES,
    DmtxSymbolSize, DmtxScheme, dmtxEncodeSetProp, dmtxEncodeDataMatrix,
    dmtxImageGetProp, dmtxEncodeCreate, dmtxEncodeDestroy
)

__all__ = [
    'decode', 'encode', 'Encoded', 'ENCODING_SCHEME_NAMES',
    'ENCODING_SIZE_NAMES', 'EXTERNAL_DEPENDENCIES',
]

ENCODING_SCHEME_PREFIX = 'DmtxScheme'
ENCODING_SIZE_PREFIX = 'DmtxSymbol'

ENCODING_SCHEME_NAMES = sorted(
    n.name[len(ENCODING_SCHEME_PREFIX):] for n in DmtxScheme
)

# Not sorting encoding size names - would need to use natural sort order;
# the existing order within DmtxSymbolSize is sensible.
ENCODING_SIZE_NAMES = [
    n.name[len(ENCODING_SIZE_PREFIX):] for n in DmtxSymbolSize
]

# A rectangle
Rect = namedtuple('Rect', 'left top width height')
Rect2 = namedtuple('Rect', 'x00 y00 x01 y01 x10 y10 x11 y11')

# Results of reading a barcode
Decoded = namedtuple('Decoded', 'data rect')

# Results of encoding data to an image
Encoded = namedtuple('Encoded', 'width height bpp pixels')

# Crude mapping from bits-per-pixels to values in DmtxPackOrder enum
_PACK_ORDER = {
    8: DmtxPackOrder.DmtxPack8bppK,
    16: DmtxPackOrder.DmtxPack16bppRGB,
    24: DmtxPackOrder.DmtxPack24bppRGB,
    32: DmtxPackOrder.DmtxPack32bppRGBX,
}


@contextmanager
def _image(pixels, width, height, pack):
    """A context manager for `DmtxImage`, created and destroyed by
    `dmtxImageCreate` and `dmtxImageDestroy`.

    Args:
        pixels (:obj:):
        width (int):
        height (int):
        pack (int):

    Yields:
        DmtxImage: The created image

    Raises:
        PyLibDMTXError: If the image could not be created.
    """
    image = dmtxImageCreate(pixels, width, height, pack)
    if not image:
        raise PyLibDMTXError('Could not create image')
    else:
        try:
            yield image
        finally:
            dmtxImageDestroy(byref(image))


@contextmanager
def _decoder(image, shrink):
    """A context manager for `DmtxDecode`, created and destroyed by
    `dmtxDecodeCreate` and `dmtxDecodeDestroy`.

    Args:
        image (POINTER(DmtxImage)):
        shrink (int):

    Yields:
        POINTER(DmtxDecode): The created decoder

    Raises:
        PyLibDMTXError: If the decoder could not be created.
    """
    decoder = dmtxDecodeCreate(image, shrink)
    if not decoder:
        raise PyLibDMTXError('Could not create decoder')
    else:
        try:
            yield decoder
        finally:
            dmtxDecodeDestroy(byref(decoder))


@contextmanager
def _region(decoder, timeout):
    """A context manager for `DmtxRegion`, created and destroyed by
    `dmtxRegionFindNext` and `dmtxRegionDestroy`.

    Args:
        decoder (POINTER(DmtxDecode)):
        timeout (int or None):

    Yields:
        DmtxRegion: The next region or None, if all regions have been found.
    """
    region = dmtxRegionFindNext(decoder, timeout)
    try:
        yield region
    finally:
        if region:
            dmtxRegionDestroy(byref(region))


@contextmanager
def _decoded_matrix_region(decoder, region, corrections):
    """A context manager for `DmtxMessage`, created and destoyed by
    `dmtxDecodeMatrixRegion` and `dmtxMessageDestroy`.

    Args:
        decoder (POINTER(DmtxDecode)):
        region (POINTER(DmtxRegion)):
        corrections (int):

    Yields:
        DmtxMessage: The message.
    """
    message = dmtxDecodeMatrixRegion(decoder, region, corrections)
    try:
        yield message
    finally:
        if message:
            dmtxMessageDestroy(byref(message))


def _decode_region(decoder, region, corrections, shrink):
    """Decodes and returns the value in a region.

    Args:
        region (DmtxRegion):

    Yields:
        Decoded or None: The decoded value.
    """
    with _decoded_matrix_region(decoder, region, corrections) as msg:
        if msg:
            # Coordinates
            p00 = DmtxVector2()
            p11 = DmtxVector2(1.0, 1.0)
            p10 = DmtxVector2(1.0, 0.0)
            p01 = DmtxVector2(0.0, 1.0)            
            dmtxMatrix3VMultiplyBy(p00, region.contents.fit2raw)
            dmtxMatrix3VMultiplyBy(p11, region.contents.fit2raw)
            dmtxMatrix3VMultiplyBy(p01, region.contents.fit2raw)
            dmtxMatrix3VMultiplyBy(p10, region.contents.fit2raw)
            x00 = int((shrink * p00.X) + 0.5)
            y00 = int((shrink * p00.Y) + 0.5)
            x11 = int((shrink * p11.X) + 0.5)
            y11 = int((shrink * p11.Y) + 0.5)
            x10 = int((shrink * p10.X) + 0.5)
            y10 = int((shrink * p10.Y) + 0.5)
            x01 = int((shrink * p01.X) + 0.5)
            y01 = int((shrink * p01.Y) + 0.5)
            return Decoded(
                string_at(msg.contents.output),
                Rect2(x00,y00,x01,y01,x10,y10,x11,y11)
            )
        else:
            return None


def _pixel_data(image):
    """Returns (pixels, width, height, bpp)

    Returns:
        :obj: `tuple` (pixels, width, height, bpp)
    """
    # Test for PIL.Image, numpy.ndarray, and imageio.core.util without
    # requiring that cv2, PIL, or imageio are installed.

    image_type = str(type(image))
    if 'PIL.' in image_type:
        pixels = image.tobytes()
        width, height = image.size
    elif 'numpy.ndarray' in image_type or 'imageio.core.util' in image_type:
        # Different versions of imageio use a subclass of numpy.ndarray
        # called either imageio.core.util.Image or imageio.core.util.Array.
        if 'uint8' != str(image.dtype):
            image = image.astype('uint8')
        try:
            pixels = image.tobytes()
        except AttributeError:
            # `numpy.ndarray.tobytes()` introduced in `numpy` 1.9.0 - use the
            # older `tostring` method.
            pixels = image.tostring()
        height, width = image.shape[:2]
    else:
        # image should be a tuple (pixels, width, height)
        pixels, width, height = image

        # Check dimensions
        if 0 != len(pixels) % (width * height):
            raise PyLibDMTXError(
                (
                    'Inconsistent dimensions: image data of {0} bytes is not '
                    'divisible by (width x height = {1})'
                ).format(len(pixels), (width * height))
            )

    # Compute bits-per-pixel
    bpp = 8 * len(pixels) // (width * height)
    if bpp not in _PACK_ORDER:
        raise PyLibDMTXError(
            'Unsupported bits-per-pixel: [{0}] Should be one of {1}'.format(
                bpp, sorted(_PACK_ORDER.keys())
            )
        )

    return pixels, width, height, bpp


def decode(image, timeout=None, gap_size=None, shrink=1, shape=None,
           deviation=None, threshold=None, min_edge=None, max_edge=None,
           corrections=None, max_count=None):
    """Decodes datamatrix barcodes in `image`.

    Args:
        image: `numpy.ndarray`, `PIL.Image` or tuple (pixels, width, height)
        timeout (int): milliseconds
        gap_size (int):
        shrink (int):
        shape (int):
        deviation (int):
        threshold (int):
        min_edge (int):
        max_edge (int):
        corrections (int):
        max_count (int): stop after reading this many barcodes. `None` to read
            as many as possible.

    Returns:
        :obj:`list` of :obj:`Decoded`: The values decoded from barcodes.
    """
    dmtx_timeout = None
    if timeout:
        now = dmtxTimeNow()
        dmtx_timeout = dmtxTimeAdd(now, timeout)

    if max_count is not None and max_count < 1:
        raise ValueError('Invalid max_count [{0}]'.format(max_count))

    pixels, width, height, bpp = _pixel_data(image)

    results = []
    with _image(
        cast(pixels, c_ubyte_p), width, height, _PACK_ORDER[bpp]
    ) as img:
        with _decoder(img, shrink) as decoder:
            properties = [
                (DmtxProperty.DmtxPropScanGap, gap_size),
                (DmtxProperty.DmtxPropSymbolSize, shape),
                (DmtxProperty.DmtxPropSquareDevn, deviation),
                (DmtxProperty.DmtxPropEdgeThresh, threshold),
                (DmtxProperty.DmtxPropEdgeMin, min_edge),
                (DmtxProperty.DmtxPropEdgeMax, max_edge)
            ]

            # Set only those properties with a non-None value
            for prop, value in ((p, v) for p, v in properties if v is not None):
                dmtxDecodeSetProp(decoder, prop, value)

            if not corrections:
                corrections = DmtxUndefined

            while True:
                with _region(decoder, dmtx_timeout) as region:
                    # Finished file or ran out of time before finding another
                    # region
                    if not region:
                        break
                    else:
                        # Decoded
                        res = _decode_region(
                            decoder, region, corrections, shrink
                        )
                        if res:
                            results.append(res)

                            # Stop if we've reached maximum count
                            if max_count and len(results) == max_count:
                                break

    return results


@contextmanager
def _encoder():
    encoder = dmtxEncodeCreate()
    if not encoder:
        raise PyLibDMTXError('Could not create encoder')

    try:
        yield encoder
    finally:
        dmtxEncodeDestroy(byref(encoder))


def encode(data, scheme=None, size=None):
    """
    Encodes `data` in a DataMatrix image.

    For now bpp is the libdmtx default which is 24

    Args:
        data: bytes instance
        scheme: encoding scheme - one of `ENCODING_SCHEME_NAMES`, or `None`.
            If `None`, defaults to 'Ascii'.
        size: image dimensions - one of `ENCODING_SIZE_NAMES`, or `None`.
            If `None`, defaults to 'ShapeAuto'.

    Returns:
        Encoded: with properties `(width, height, bpp, pixels)`.
        You can use that result to build a PIL image:

            Image.frombytes('RGB', (width, height), pixels)

    """

    size = size if size else 'ShapeAuto'
    size_name = '{0}{1}'.format(ENCODING_SIZE_PREFIX, size)
    if not hasattr(DmtxSymbolSize, size_name):
        raise PyLibDMTXError(
            'Invalid size [{0}]: should be one of {1}'.format(
                size, ENCODING_SIZE_NAMES
            )
        )
    size = getattr(DmtxSymbolSize, size_name)

    scheme = scheme if scheme else 'Ascii'
    scheme_name = '{0}{1}'.format(
        ENCODING_SCHEME_PREFIX, scheme.capitalize()
    )
    if not hasattr(DmtxScheme, scheme_name):
        raise PyLibDMTXError(
            'Invalid scheme [{0}]: should be one of {1}'.format(
                scheme, ENCODING_SCHEME_NAMES
            )
        )
    scheme = getattr(DmtxScheme, scheme_name)

    with _encoder() as encoder:
        dmtxEncodeSetProp(encoder, DmtxProperty.DmtxPropScheme, scheme)
        dmtxEncodeSetProp(encoder, DmtxProperty.DmtxPropSizeRequest, size)

        if dmtxEncodeDataMatrix(encoder, len(data), cast(data, c_ubyte_p)) == 0:
            raise PyLibDMTXError(
                'Could not encode data, possibly because the image is not '
                'large enough to contain the data'
            )

        w, h, bpp = map(
            partial(dmtxImageGetProp, encoder[0].image),
            (
                DmtxProperty.DmtxPropWidth, DmtxProperty.DmtxPropHeight,
                DmtxProperty.DmtxPropBitsPerPixel
            )
        )
        size = w * h * bpp // 8
        pixels = cast(
            encoder[0].image[0].pxl, ctypes.POINTER(ctypes.c_ubyte * size)
        )
        return Encoded(
            width=w, height=h, bpp=bpp, pixels=ctypes.string_at(pixels, size)
        )
